src/bsoncxx/test/bson_builder.cpp

```diff
@@ -1234,4 +1246,447 @@ TEST_CASE("builder::stream::document throws on consecutive keys", "[bsoncxx::bui
+    bson_t expected;
+    bson_init(&expected);
+
+    bson_append_int64(&expected, "foo", -1, 100);
+
+    SECTION("raw int64") {
+        bson b{"foo", std::int64_t(100)};
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    SECTION("b_int64") {
+        bson b{"foo", types::b_int64{100}};
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    bson_destroy(&expected);
+}
+
+TEST_CASE("bson builder appends decimal128", "[bsoncxx::builder::bson]") {
+    bson_t expected;
+    bson_init(&expected);
+
+    bson_decimal128_t d128;
+    bson_decimal128_from_string("-1234E+999", &d128);
+    bson_append_decimal128(&expected, "foo", -1, &d128);
+
+    SECTION("b_decimal128 works") {
+        bson b{"foo", types::b_decimal128{"-1234E+999"}};
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    SECTION("raw bsoncxx::decimal128 works") {
+        bson b{"foo", bsoncxx::decimal128{"-1234E+999"}};
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    SECTION("bsoncxx::types::bson_value::view with decimal128 works") {
+        auto d = types::b_decimal128{"-1234E+999"};
+        auto v = types::bson_value::view{d};
+
+        REQUIRE(v.get_decimal128() == d);
+
+        bson b{"foo", v};
+        bson_eq_document(&expected, b.view().get_document());
+    }
+
+    bson_destroy(&expected);
+}
+
+TEST_CASE("bson builder appends minkey", "[bsoncxx::builder::bson]") {
+    bson_t expected;
+    bson_init(&expected);
+
+    bson_append_minkey(&expected, "foo", -1);
+
+    bson b{"foo", types::b_minkey{}};
+    bson_eq_document(&expected, b.view().get_document());
+
+    bson_destroy(&expected);
+}
+
+TEST_CASE("bson builder appends maxkey", "[bsoncxx::builder::bson]") {
+    bson_t expected;
+    bson_init(&expected);
+
+    bson_append_maxkey(&expected, "foo", -1);
+
+    bson b{"foo", types::b_maxkey{}};
+    bson_eq_document(&expected, b.view().get_document());
+
+    bson_destroy(&expected);
+}
+
+TEST_CASE("bson builder appends array", "[bsoncxx::builder::bson]") {
+    bson_t expected, child;
+    bson_init(&expected);
+    bson_init(&child);
+
+    bson_append_utf8(&child, "0", -1, "bar", -1);
+    bson_append_array(&expected, "foo", -1, &child);
+
+    bson arr{"bar"};
+
+    bson b{"foo", arr};
+    bson_eq_document(&expected, b.view().get_document());
+
+    bson_destroy(&expected);
+    bson_destroy(&child);
+}
+
+TEST_CASE("bson builder appends document", "[bsoncxx::builder::bson]") {
+    bson_t expected, child;
+    bson_init(&expected);
+    bson_init(&child);
+
+    bson_append_utf8(&child, "bar", -1, "baz", -1);
+    bson_append_document(&expected, "foo", -1, &child);
+
+    bson doc{"bar", "baz"};
+
+    bson b{"foo", doc};
+    bson_eq_document(&expected, b.view().get_document());
+
+    bson_destroy(&expected);
+    bson_destroy(&child);
+}
+
+TEST_CASE("bson builder appends inline array", "[bsoncxx::builder::bson]") {
```

Comment:
I added a way for users to specify if they'd want to forgo type deduction altogether.

Reply:
I think that's a good change!  However I think the syntax for using it could still be improved:

`builder::list b{"fruits", builder::list({"mango", "apple"}, false, true)};`

It's a little long, and possibly confusing to people to be thinking about the recursive nature of the builder::list constructor, and having true/false values floating around will probably confuse people, or have them looking back at the documentation constantly.  Having to create the list builder with those boolean parameters in-line also breaks the JSON syntax look of this constructor.

I propose adding a way to have the type selection look a bit more like a type cast, which is shorter, avoids the true/false parameters, and doesn't break the JSON-esque syntax flow as much:

```
builder::list b{"fruits", list::array{"mango", "apple"}};
builder::list b{"fruits", list::document{"mango", "apple"}};
```

I noodled around with this on a branch, and you can make it work by adding two little converter classes as subclasses in the list builder:

```
namespace builder {

class list {
public:
   /// constructors, etc

   class array {
   public:
      // default ctor supports empty arrays
      array() {}

      // regular ctor stores the init list so we can use it to convert to a list builder later
      array(std::initializer_list<list> init) : _init(std::move(init)) {}

      // implicit conversion operator to make this usable inline
      operator list() const { 
         return list(_init, false, true);
      }

   private:
      std::initializer_list<list> _init;
   };

   class document {
      // same as "array" subclass, but operator function creates builder::list with false, false
   };
};
```

Thoughts?

Reply:
Oh, that's a lot nicer. I added those classes and changed a few of the tests to use it. I did, though, leave all of the tests for the `type_deduction` and `is_array` booleans since those are still used under the hood.